### DISCUSSION TITLE
srsRAN Project does not utilize the whole CPU

### DISCUSSION THREAD

**Discussion by oguztudo** [srsran/srsRAN_Project#122] [General Help]
Hello everybody,

as the title says the gNB does not use all of the available CPU and there has massive dropoffs in throughput on higher bandwidths (starting at 70 MHz). 
I am using 92.16 MHz as the sample rate starting at 50 MHz. As expected an increase in BW leads to an increase in the CPU-usage but only two cores can reach the maximum power while the rest peaks at 60% usage. Also the number of underruns increase with higher Bandwidths as well.
Are there any ways i can utilize the unused CPU?

I am already running the performance script which is coming with srsRAN Project.
If there are any special logs or pcaps to be provided feel free to ask for them.

**Comment by ismagom** [comment]
To run these high BW and if you have more than 4 cores in your CPU, use this option: 

```
ru_sdr:
  expert_cfg:
    low_phy_thread_profile: quad
```

If the problem persists, can you paste here the output of the console trace please?

**Reply by oguztudo** [reply]
Hello Ismael, 
the error persists unfortunately. I added an example of the trace down below. It is not a long excerpt but in my opinion enough to show the general behavior.

[quad_log.txt](https://github.com/srsran/srsRAN_Project/files/12014151/quad_log.txt)

Here is also the CPU-usage. The gNB starts where the few spikes occur. The DL Speedtest starts when the red Graph rises and the DL is when the green graph rises. You can see that in DL there's only one core that is being fully used while in the UL two cores split up the work and switch back and forth. I did not expect the whole 16 cores to be used but i can't even reach 4 in the quad executor mode.

![quad](https://github.com/srsran/srsRAN_Project/assets/129198919/e2caa476-d49b-4446-92cb-e56f78952959)

In addition to the internet speedtest i also did the iperf3 in udp transmission mode for the DL. Here the trace gives me constant values of around 230M for dual and 110M for quad executor mode. Even in single mode i can get at around 205M +- 15M (tolerance is higher here).

**Reply by ismagom** [reply]
Can you attach the log file please?

**Reply by oguztudo** [reply]
Of course, I had to shorten it a bit to match the 25MB upload limit.

[gnb.log](https://github.com/srsran/srsRAN_Project/files/12014450/gnb.log)

**Reply by ismagom** [reply]
Thanks for the log. Unfortunately in the attached log the BLER is quite low :) However, I do can see some warning due to late PHY processing, such as: 

```
2023-07-11T12:03:59.942861 [Low-PHY ] [W] Real-time failure in low-phy: Downlink data late for sector 0 and slot 337.5.
2023-07-11T12:03:59.956178 [Low-PHY ] [W] Real-time failure in low-phy: Downlink data late for sector 0 and slot 338.12.
2023-07-11T12:03:59.956749 [Low-PHY ] [W] Real-time failure in low-phy: Downlink data late for sector 0 and slot 338.13.
2023-07-11T12:03:59.957282 [Low-PHY ] [W] Real-time failure in low-phy: Downlink data late for sector 0 and slot 338.14.
2023-07-11T12:03:59.957808 [Low-PHY ] [W] Real-time failure in low-phy: Downlink data late for sector 0 and slot 338.15.
2023-07-11T12:03:59.962783 [Low-PHY ] [W] Real-time failure in low-phy: Downlink data late for sector 0 and slot 339.5.
2023-07-11T12:04:00.206645 [Low-PHY ] [W] Real-time failure in low-phy: Downlink data late for sector 0 and slot 363.13.
2023-07-11T12:04:00.207192 [Low-PHY ] [W] Real-time failure in low-phy: Downlink data late for sector 0 and slot 363.14.
```

These could cause BLER. One thing you can try is these expert parameters (top level in the configuration file)

```
expert_phy: 
  nof_ul_threads: 3
  nof_pdsch_threads: 3
```
Can you use these parameters and paste the console trace here again? 

Also, in terms of using the CPU cores, I don't think it's a metric you need to look at. It's more important to look at the total CPU usage of the software. It seems you are running 8 "true" CPU cores. With this you should probably be able to reach 400-600% total CPU usage and still have a stable connection.

**Reply by oguztudo** [reply]
of course!

Here is the trace:
[quad_log_w_expert_phy.txt](https://github.com/srsran/srsRAN_Project/files/12028417/quad_log_w_expert_phy.txt)

and also the log:
[gnb_ex_phy.log](https://github.com/srsran/srsRAN_Project/files/12028435/gnb_ex_phy.log)


Now i got the issue that my device disconnects during the upload more frequently with the following message appearing in the tail of the amf log:

07/12 15:13:36.050: [amf] INFO: UE Context Release [Action:2] (../src/amf/ngap-handler.c:1632)
07/12 15:13:36.050: [amf] INFO:     RAN_UE_NGAP_ID[0] AMF_UE_NGAP_ID[36] (../src/amf/ngap-handler.c:1633)
07/12 15:13:36.050: [amf] INFO:     SUCI[suci-0-001-01-0-0-0-0000082271] (../src/amf/ngap-handler.c:1636)
07/12 15:13:36.050: [amf] INFO: [Removed] Number of gNB-UEs is now 1 (../src/amf/context.c:2509)

**Reply by ismagom** [reply]
Can you share the trace when you use dual mode and also the model of your CPU? thanks

**Reply by oguztudo** [reply]
should i keep the expert_phy parameters active while doing so?

**Reply by ismagom** [reply]
yes please

**Reply by oguztudo** [reply]
Here is the trace for three consecutive speetdests using dual mode.

[dual_log_w_expert_phy.txt](https://github.com/srsran/srsRAN_Project/files/12029468/dual_log_w_expert_phy.txt)

The CPU i am using is an Intel Xeon W-2245.

**Reply by ismagom** [reply]
Thanks. Can you share the full config file for reference?

**Reply by oguztudo** [reply]
Sorry for the late reply, here is the configuration:

[gnb.txt](https://github.com/srsran/srsRAN_Project/files/12049776/gnb.txt)

**Reply by ismagom** [reply]
Hi @oguztudo , I'm sorry I tried to reproduce with a similar CPU (Xeon Gold 5317) but I couldn't. Could you run another test with quad mode, this time doing UDP DL iperf2 test only and send logs and console trace please? 

Thank you very much

**Reply by oguztudo** [reply]
Hello @ismagom and thanks testing the configtuation. 
I can do that but not before wednesday unfortunately...

**Reply by ismagom** [reply]
One more thing, with the X3x0 there is sometimes a lot of LO leakage. Can you try using the option: 

```
ru_sdr: 
    lo_offset: 40
```

to move LO away of the band?

**Reply by oguztudo** [reply]
Hello @ismagom,
The offset of the local oscillator was something i already treid but it did not solve the issue.
I was working with srsRAN during my thesis which i finished about 3 weeks ago, but i was hoping to be able to do some further testing on the day where i packed all of my stuff. Unfortunately this was not possible and therefore I think this issue can be closed...
